### DISCUSSION TITLE
Modbus Error: ... read of 8 bytes without response from unit before it closed connection

### DISCUSSION THREAD

**Discussion by rdorsch** [pymodbus-dev/pymodbus#1457] [Help]
Hello,

I have a strange observation here:

I do a periodic read_holding_registers() operation. If I read every 10 secs, everything works super reliable, if I read every 60 secs, I get a connection error. The server is an Alfen Eve NG910 wallbox:

Example code:

```
rd@evcc:~$ cat alfen-test-minimal.py
import pymodbus.payload
import pymodbus.client
import time
import datetime
import random
    
# establish wb connection
client = pymodbus.client.ModbusTcpClient('192.168.2.231', port=502, timeout=10, auto_open=True, auto_close=False)
client.connect()

while True:
    print(datetime.datetime.now())
    try: 
        rr = client.read_holding_registers(344,count=2,slave=1)
        decoder = pymodbus.payload.BinaryPayloadDecoder.fromRegisters(rr.registers, byteorder=pymodbus.constants.Endian.Big, wordorder=pymodbus.constants.Endian.Big)
        energy_measurements = {
            'real_power': decoder.decode_32bit_float(),
        }
        r=energy_measurements["real_power"]
        print("wb_real_power: ", r)
    except Exception as e:
        print('wb_real_power_sum_exception: ')
        print(e)

    time.sleep(60)

                    
rd@evcc:~$ 
```

With the 60 seconds sleep interval, I get as output:

```
rd@evcc:~$ python3 alfen-test-minimal.py 
2023-03-26 12:07:03.655449
wb_real_power:  0.0
2023-03-26 12:08:03.784190
wb_real_power_sum_exception: 
Modbus Error: [Connection] ModbusTcpClient(192.168.2.231:502): Connection unexpectedly closed 0.00012636184692382812 seconds into read of 8 bytes without response from unit before it closed connection

```

with a 10 seconds sleep interval everything works as expected.

The issue goes away, if I add a client.connect() before the client.read_holding_registers() and a client.close() thereafter.

I assume that the wallbox closes the connection after a certain time of inactivity.

Isn't the auto_open=True supposed to automatically open the connection again?

Thanks
Rainer

**Comment by janiversen** [comment]
Where did you see auto_open it is not in our documentation.

But yes your problem is that wallbox (like many other devices) close the connection after 30 seconds.

**Comment by rdorsch** [comment]
Hi Jan,

many thanks for the quick response.

Am Sonntag, 26. März 2023, 18:13:08 CEST schrieb jan iversen:
> Where did you see auto_open it is not in our documentation.

It seems I used the wrong documentation without noticing it:

https://pymodbustcp.readthedocs.io/en/latest/quickstart/index.html

I am surprised that python does not complain about the undefined arguments.

> But yes your problem is that wallbox (like many other devices) close the
> connection after 30 seconds.

Would you open for each operation a connect() and close() if there is a risk, 
that the timeout becomes bigger than 30 seconds?

        client.connect()
        rr = client.read_holding_registers(344,count=2,slave=1)
        client.close()

Thanks
Rainer

-- 
Rainer Dorsch
http://bokomoko.de/

**Reply by janiversen** [reply]
That is clearly a legal way of doing it, On my production system I make a call every 10 seconds reading a sensor not very important but it keeps the connection open.
